[FEAT] 실시간 인기글 조회 기능 구현

Kim-TaeUk opened the pull request:
## Related Issue
<!-- feat/#issue -> dev와 같이 반영 브랜치를 표시 -->
<!-- closed #issue로 merge되면 issue가 자동으로 close되게 -->
- feat/#116 -> dev
- close #116

## Key Changes
<!-- 최대한 자세히 -->
- 실시간 인기글 조회
  - 비회원과 회원 분기 처리했습니다. 
  - 차단관계를 고려하여 피드를 조회할 때 JOIN을 쓸까 서브쿼리를 쓸까 고민하다가, QueryDsl을 이용해서 union으로 차단관계에 있는 ID를 얻고, IN 절 서브쿼리로 처리하는 방식을 사용했습니다. 
  - 비회원인 경우 차단관계를 고려하지 않아도 되기 때문에 간단하게 상위 9개 레코드를 반환하는 JPA 쿼리 메서드를 호출하도록 했습니다.

## References
<!-- 참고한 자료-->
- https://github.com/querydsl/querydsl/blob/master/querydsl-sql/src/test/java/com/querydsl/sql/UnionBase.java

Kim-TaeUk reviewed (approved):
Approve!

rinarina0429 reviewed (approved):
수고하셨습니다~!

Kim-TaeUk committed:
[FEAT] 실시간 인기글 조회를 위한 end-point 추가

Kim-TaeUk committed:
[FIX] PopularFeedRepository에 누락된 @Repository 어노테이션 추가

Kim-TaeUk committed:
[FEAT] 실시간 인기글 조회 DTO 추가

Kim-TaeUk committed:
[FEAT] PopularFeedRepository에 실시간 인기글 조회 쿼리메서드 추가

PopularFeedRepository에서 최근 인기글 상위 9개 반환

Kim-TaeUk committed:
[FEAT] 실시간 인기글 조회 서비스 로직 추가

Kim-TaeUk committed:
[REFACTOR] 인기글 조회 로직 리팩터링

method extract

Kim-TaeUk committed:
[FIX] 인기글 조회의 잘못된 end point 수정

/novels/popular -> /feeds/popular

Kim-TaeUk committed:
[FIX] 인기글 조회 로직 QueryDsl로 수정

차단 관계를 고려햐여 PopularFeedRepository에서 최근 인기글 상위 9개 반환

Kim-TaeUk committed:
[FIX] 엔드포인트에서 비회원/회원 분기 처리

비회원인 경우 principal을 null로 처리하여 분기

Kim-TaeUk committed:
[FIX] 차단 관계를 고려한 리포지토리 메서드 사용하도록 변경

Kim-TaeUk committed:
[FIX] 비회원 분기 처리

Kim-TaeUk committed:
[FEAT] white list에 인기글 조회 엔드포인트 추가

Kim-TaeUk committed:
[STYLE] private 메서드 순서 조정

메서드에서 사용되는 순서대로 순서 조정

ChaeAg commented on src/main/java/org/websoso/WSSServer/service/PopularFeedService.java:
```diff
@@ -20,4 +24,34 @@ public void createPopularFeed(Feed feed) {
+                        popularFeed.getFeed().getLikes().size(),
+                        popularFeed.getFeed().getComments().size(),
+                        popularFeed.getFeed().getIsSpoiler()
+                ))
```
P5; 오 혹시 이 DTO는 static 생성 메서드는 구현 안하신 이유가 따로 있으신가용?

Kim-TaeUk replied:
오 딱히 이유는 없었습니다.
요구사항에 맞게 작업해주는게 아니니까 dto에서 하는게 더 낫겠네요! 감사합니다!!

ChaeAg commented on src/main/java/org/websoso/WSSServer/repository/PopularFeedCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final JPAQueryFactory jpaQueryFactory;
+
+    @Override
+    public List<PopularFeed> findTodayPopularFeeds(Long userId) {
+        List<Long> blockingIds = jpaQueryFactory
+                .select(block.blockedId)
+                .from(block)
+                .where(block.blockingId.eq(userId))
+                .fetch();
+
+        List<Long> blockedIds = jpaQueryFactory
+                .select(block.blockingId)
+                .from(block)
+                .where(block.blockedId.eq(userId))
+                .fetch();
+
+        List<Long> blockIds = Stream.concat(blockingIds.stream(), blockedIds.stream())
+                .distinct()
+                .toList();
+
+        return jpaQueryFactory
+                .selectFrom(popularFeed)
+                .where(popularFeed.feed.user.userId.notIn(blockIds))
+                .orderBy(popularFeed.popularFeedId.desc())
+                .limit(9)
+                .fetch();
```
P5; 우와 대박!

Kim-TaeUk replied:
JPQL이랑 QueryDSL 둘 다 UNION을 지원하지 않아서 native Query를 쓸까 하다가 
native query가 dbms에 종속적이기도 하고(물론 dbms를 바꿀 것 같진 않긴 하지만,,),
타입 안정성을 챙길 수 없어서 QueryDSL을 선택했습니다.

ChaeAg reviewed (approved):
수고하셨습니다~~!

Kim-TaeUk committed:
[FEAT] PopularFeedGetResponse에 정적 팩터리 메서드 of 추가

Kim-TaeUk committed:
[REFACTOR] PopularFeedGetResponse 생성에 Method Reference 사용

rinarina0429 commented on src/main/java/org/websoso/WSSServer/controller/FeedController.java:
```diff
@@ -97,4 +100,16 @@ public ResponseEntity<FeedGetResponse> getFeed(Principal principal,
+        return ResponseEntity
+                .status(OK)
+                .body(popularFeedService.getPopularFeeds(user));
+    }
```
p3; 혹시 이런식으로 바꾸는건 어떠실련지요? 분기처리를 해야 하는 다른 로직들(FeedController, UserNovelController 등)에서 위와 같이 쓰고 있어 제안해봅니다!
```suggestion
    @GetMapping("/popular")
    public ResponseEntity<PopularFeedsGetResponse> getPopularFeeds(Principal principal) {
        User user = principal == null ? null : userService.getUserOrException(Long.valueOf(principal.getName()));
        return ResponseEntity
                .status(OK)
                .body(popularFeedService.getPopularFeeds(user));
    }
```

Kim-TaeUk replied:
오호 삼항 연산자를 통해 간소화하는 코드네요
반영하겠습니다!

rinarina0429 commented on src/main/java/org/websoso/WSSServer/service/PopularFeedService.java:
```diff
@@ -20,4 +24,28 @@ public void createPopularFeed(Feed feed) {
+        return popularFeedRepository.findTodayPopularFeeds(user.getUserId());
+    }
+
+    private static List<PopularFeedGetResponse> getPopularFeedGetResponses(List<PopularFeed> popularFeeds) {
```
p5; get 말고 convertToPopularFeedGetResponse나 convertToDto와 같은 이름은 어떨까요?

Kim-TaeUk replied:
`mapToPopularFeedGetResponseList`로 rename했습니다!

Kim-TaeUk replied:
get해오는게 아니니까 매우 안좋은 네이밍이었네요! 감사합니다.

Kim-TaeUk replied:
메서드 네이밍 관련해서 컨벤션으로 정해보면 어떨까 생각들었던게 있어서 공유합니다!

Repository layer에서 생성 - save, 조회 - find
Service layer에서 생성 - create, 조회 - get
으로 사용하시는거 어떤가요??

@rinarina0429 @ChaeAg

rinarina0429 replied:
너무 조와요. 앞으로 저렇게 네이밍 하도록 하겟숨댜

rinarina0429 replied:
채은언니 확인을 위해 resolve하지 않고 남겨두겠습니다! 머지 전에 resolve 해주세용

rinarina0429 commented on src/main/java/org/websoso/WSSServer/service/PopularFeedService.java:
```diff
@@ -20,4 +24,28 @@ public void createPopularFeed(Feed feed) {
+    public PopularFeedsGetResponse getPopularFeeds(User user) {
+        List<PopularFeed> popularFeeds = findPopularFeeds(user);
+        List<PopularFeedGetResponse> popularFeedGetResponses = getPopularFeedGetResponses(popularFeeds);
+        return getPopularFeedsGetResponse(popularFeedGetResponses);
```
p5; 요 부분은 그냥 함수 사용하지 않고 그대로 쓰는게 더 잘 읽히지 않을까요?
```suggestion
        return new PopularFeedsGetResponse(popularFeedGetResponses);
```

Kim-TaeUk replied:
오호 메서드 추출하다가 잘못 들어간 것 같습니다.
반영했습니다!

rinarina0429 commented on src/main/java/org/websoso/WSSServer/repository/PopularFeedCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public List<PopularFeed> findTodayPopularFeeds(Long userId) {
+        List<Long> blockingIds = jpaQueryFactory
+                .select(block.blockedId)
+                .from(block)
+                .where(block.blockingId.eq(userId))
+                .fetch();
+
+        List<Long> blockedIds = jpaQueryFactory
+                .select(block.blockingId)
+                .from(block)
+                .where(block.blockedId.eq(userId))
+                .fetch();
+
+        List<Long> blockIds = Stream.concat(blockingIds.stream(), blockedIds.stream())
+                .distinct()
+                .toList();
```
q; 혹시! 이 세 친구들을 하나로 묶을 수는 없을까요? 요롷게는 안되려나요
```suggestion
private List<Long> findBlockedOrBlockingUserIds(Long userId) {
        return jpaQueryFactory
                .select(block.blockingId)
                .from(block)
                .where(block.blockingId.eq(userId)
                        .or(block.blockedId.eq(userId)))
                .fetch();
    }
```

Kim-TaeUk replied:
로직이 달라질 것 같은데요~!

rinarina0429 replied:
아니 그렇네요! 왜 저렇게 썼었지 제가 재시도를 해보겠습니다
```suggestion
List<Long> blockIds = jpaQueryFactory
        .select(block.blockingId)
        .from(block)
        .where(block.blockingId.eq(userId)
            .or(block.blockedId.eq(userId)))
        .union(
            jpaQueryFactory
            .select(block.blockedId)
            .from(block)
            .where(block.blockingId.eq(userId)
                .or(block.blockedId.eq(userId)))
        )
        .fetch();
```
이렇게는 어떠신가요? union으로 합치게 되면 distinct로 자동으로 적용되고 별개로 `Stream.concat()`같은 과정도 거칠 필요가 없어질 것 같아서요!

Kim-TaeUk replied:
`JPQL`이랑 `QueryDSL`에서는 UNION을 지원하지 않는 것으로 알고 있는데욥, JPAQueryFactory로 .union()을 사용하는 방법 레퍼런스 참고하신 거 있나욥!?

---

query-dsl 공식 github discussions에도 2021년 글이긴 한데 jpa에서는 지원하지 않는다고 나와있고, (https://github.com/querydsl/querydsl/discussions/2927)

query-dsl 공식 github에 jpa가 아닌 native sql로 사용하는 방식은 지원하고 있는 것 같습니다. -> com.querydsl.jpa가 아닌 com.querydsl.sql의 SQLQueryFactory를 사용(https://github.com/querydsl/querydsl/blob/8f96f416270d0353f90a6551547906f3c217833a/querydsl-sql/src/test/java/com/querydsl/sql/UnionBase.java#L73)

---

> JPQL이랑 QueryDSL 둘 다 UNION을 지원하지 않아서 native Query를 쓸까 하다가
native query가 dbms에 종속적이기도 하고(물론 dbms를 바꿀 것 같진 않긴 하지만,,),
타입 안정성을 챙길 수 없어서 QueryDSL을 선택했습니다.

SQLQueryFactory를 사용하고 SQLTemplates로 DBMS에 맞는 SQL 문법을 적용해서 사용할 수 있긴 할 것 같아요!

rinarina0429 replied:
아 그렇네요 그냥 union을 쓰면 편할 것 같다는 생각이었는데

> JPQL이랑 QueryDSL 둘 다 UNION을 지원하지 않아서 native Query를 쓸까 하다가

이걸 읽고 1초만에 까먹은 금붕어가 되었습니다.. 일단은 이대로 사용하는 것도 좋을 것 같네요!

rinarina0429 reviewed (changes requested):
수고하셨습니다!! 리뷰 한번씩만 확인 부탁드려요~

Kim-TaeUk committed:
[REFACTOR] 엔드포인트 비회원/회원 분기 처리 삼항 연산자로 리팩터링

Kim-TaeUk committed:
[REMOVE] 가독성을 떨어뜨리는 getPopularFeedsGetResponse 메서드 삭제

Kim-TaeUk committed:
[RENAME] 의도를 알기 힘든 메서드 rename

List<PopularFeed>를 List<PopularFeedGetResponse>로 map하는 메서드의 이름 rename
: getPopularFeedGetResponses -> mapToPopularFeedGetResponseList

rinarina0429 reviewed (approved):
수고하셨습니다~~

Kim-TaeUk committed:
Merge branch 'dev' into feat/#116
